src/platform/qt/scripting/ScriptingTextBuffer.cpp

```diff
@@ -49,26 +49,45 @@ void ScriptingTextBuffer::setBufferName(const QString& name) {
 	emit bufferNameChanged(name);
 }
 
+void ScriptingTextBuffer::lineBreak() {
+	bool nextBlockExists = m_shim.cursor.movePosition(QTextCursor::NextBlock, QTextCursor::MoveAnchor, 1);
+	if (!nextBlockExists) {
+		m_shim.cursor.movePosition(QTextCursor::EndOfBlock, QTextCursor::MoveAnchor, 1);
+		m_shim.cursor.insertBlock();
+	}
+}
+
 void ScriptingTextBuffer::print(const QString& text) {
+	static const int tabStop = 4;
 	QMutexLocker locker(&m_mutex);
-	for (QString split : text.split('\n')) {
-		while (m_shim.cursor.positionInBlock() + split.length() > m_dims.width()) {
-			int cut = m_dims.width() - m_shim.cursor.positionInBlock();
-			if (!m_shim.cursor.atBlockEnd()) {
-				m_shim.cursor.movePosition(QTextCursor::EndOfBlock, QTextCursor::KeepAnchor);
-			}
-			m_shim.cursor.insertText(split.left(cut));
-			if (m_shim.cursor.atEnd()) {
-				m_shim.cursor.insertBlock();
+
+	QTextCursor& cursor = m_shim.cursor;
+	int lineWidth = m_dims.width();
+
+	for (const QChar& ch : text) {
+		if (ch == '\t') {
+			int column = cursor.positionInBlock();
+			int move = tabStop - (column % tabStop) + 1;
+			if (column + move >= lineWidth) {
+				lineBreak();
+			} else if (column + move <= cursor.block().length()) {
+				cursor.movePosition(QTextCursor::Right, QTextCursor::MoveAnchor, move - 1);
 			} else {
-				m_shim.cursor.movePosition(QTextCursor::NextBlock, QTextCursor::MoveAnchor, 1);
+				move = column + move - cursor.block().length();
+				cursor.movePosition(QTextCursor::EndOfLine, QTextCursor::MoveAnchor, 1);
+				cursor.insertText(QString(move, ' '));
 			}
-			split = split.mid(cut);
-		}
-		if (!m_shim.cursor.atBlockEnd()) {
-			m_shim.cursor.movePosition(QTextCursor::NextCharacter, QTextCursor::KeepAnchor, split.length());
+		} else if (ch == '\n') {
+			lineBreak();
+		} else if (ch == '\r') {
+			cursor.movePosition(QTextCursor::StartOfLine, QTextCursor::MoveAnchor, 1);
+		} else {
+			if (cursor.positionInBlock() >= lineWidth) {
+				lineBreak();
+			}
+			cursor.movePosition(QTextCursor::Right, QTextCursor::KeepAnchor, 1);
+			cursor.insertText(ch);
```

Comment:
Inserting characters one byte at a time seems...potentially slow. Have you made sure that this doesn't cause performance issues?

Reply:
Refactored. Performance delta against master is less than 10% (compared to 1000% from the old version).